### PR TITLE
Convert RulesDeploy.js into TypeScript

### PR DESCRIPTION
**Pull request by ryanpbrewster**
I tried to keep this as unobtrusive as possible. There's a lot of line-noise involved in converting JS to TypeScript just due to eslint, so my preference would be to merge this as-is and then "upgrade" the typescript to be more idiomatic in a separate PR, where the style changes can be inspected alone.

### CODE REVIEW

File: src/RulesDeploy.ts [hunk 1]
```diff
@@ -33,46 +38,46 @@ RulesDeploy.prototype = {
...
   /**
    * Compile all rulesets tied to this deploy, rejecting on first
    * compilation error.
    */
-  compile: function() {
-    var self = this;
-    var promises = [];
-    _.forEach(this.rulesFiles, function(files, filename) {
+  compile(): Promise<any> {
+    const self = this;
```
Comment (bkendall) [inline src/RulesDeploy.ts]: Throughout the file: `self` can be removed (and uses of `self` can be replaced with `this`). The references should all work fine.

File: src/RulesDeploy.ts [hunk 2]
```diff
@@ -81,22 +86,22 @@ RulesDeploy.prototype = {
             clc.bold(resourceName)
         );
       });
-  },
+  }
 
-  _compileRuleset: function(filename, files) {
+  _compileRuleset(filename: any, files: any): Promise<any> {
```
Comment (bkendall) [inline src/RulesDeploy.ts]: `_` is an indicator of `private`, so this should be `private compileRuleset` (and any calls to it adjusted)

File: src/RulesDeploy.ts [hunk 3]
```diff
@@ -81,22 +86,22 @@ RulesDeploy.prototype = {
...
-    var self = this;
-    return gcp.rules.testRuleset(self.options.project, files).then(function(response) {
+    const self = this;
+    return gcp.rules.testRuleset(self.options.project, files).then((response: any) => {
       if (response.body && response.body.issues && response.body.issues.length > 0) {
-        var warnings = [];
-        var errors = [];
-        response.body.issues.forEach(function(issue) {
-          var issueMessage =
+        const warnings: any = [];
```
Comment (bkendall) [inline src/RulesDeploy.ts]: These are really `[]any`, right? (`warnings` and `errors`)

File: src/RulesDeploy.ts [hunk 4]
```diff
@@ -135,7 +140,7 @@ RulesDeploy.prototype = {
       );
       return Promise.resolve();
     });
-  },
-};
+  }
+}
 
-module.exports = RulesDeploy;
+export = RulesDeploy;
```
Comment (bkendall) [inline src/RulesDeploy.ts]: with the `export` keyword in the `class` declaration, this line won't be necessary.

File: src/RulesDeploy.ts [hunk 5]
```diff
@@ -1,30 +1,35 @@
...
+import utils = require("./utils");
 
-function RulesDeploy(options, type) {
-  this.type = type;
-  this.options = options;
-  this.project = options.project;
-  this.rulesFiles = {};
-  this.rulesetNames = {};
-}
+class RulesDeploy {
```
[the remark this answers is not in the document]
Comment (ryanpbrewster) [reply src/RulesDeploy.ts]: Indeed, that was my first attempt and it broke all the imports. How do I fix those?

They currently look like
```
var RulesDeploy = require("../RulesDeploy");
```

Maybe that turns into
```
var RulesDeploy = require("../RulesDeploy").RulesDeploy;
```
?

File: src/RulesDeploy.ts [hunk 6]
```diff
@@ -1,30 +1,35 @@
 "use strict";
```
[the remark this answers is not in the document]
Comment (ryanpbrewster) [reply src/RulesDeploy.ts]: done

File: src/RulesDeploy.ts (hunk 1 above)
Comment (ryanpbrewster) [reply src/RulesDeploy.ts]: done

File: src/RulesDeploy.ts (hunk 2 above)
Comment (ryanpbrewster) [reply src/RulesDeploy.ts]: done

File: src/RulesDeploy.ts (hunk 3 above)
Comment (ryanpbrewster) [reply src/RulesDeploy.ts]: done

File: src/RulesDeploy.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (bkendall) [reply src/RulesDeploy.ts]:
```typescript
const { RulesDeploy } = require("../RulesDeploy");
```

[the remark this answers is not in the document]
Comment (bkendall) [reply src/RulesDeploy.ts]: (That's assuming `export class RulesDeploy`)

File: src/RulesDeploy.ts (hunk 4 above)
Comment (ryanpbrewster) [reply src/RulesDeploy.ts]: done

**Pull request merged**